### ISSUE TITLE
[BUG/ISSUE] Memory leaks in HEMCO standalone

### ISSUE THREAD

**Issue by msulprizio** [geoschem/HEMCO#190]
I noticed there might be memory leaks in the HEMCO standalone simulation. When running high-resolution (0.25x0.3125), my standalone simulations would eventually run out of memory no matter how much I increased the memory. 

In an initial attempt to diagnose this I ran with the new sanitize option added in https://github.com/geoschem/HEMCO/pull/159. I ran a 1-month 4x5 MERRA-2 HEMCO standalone simulation with the 14.1.0 HEMCO_Config.rc. My log file is 
[attached](https://github.com/geoschem/HEMCO/files/10702329/slurm-41751397.txt).

**Comment by msulprizio** [comment]
Tagging @yantosca

**Comment by yantosca** [comment]
I found this on the web: http://www.personal.psu.edu/jhm/f90/statements/nullify.html

TL;DR, when you ALLOCATE and immediately NULLIFY, that makes blocks of memory inaccessible. This was being done in routine `ReadList_Init`.  Working on a fix.

**Comment by yantosca** [comment]
There are also memory leaks reported in filedata_mod and hcox_dustdead_mod.  Looking at these.

**Comment by lizziel** [comment]
It looks like the biggest leak by many orders of magnitude is this one:
```
Direct leak of 926512128 byte(s) in 744 object(s) allocated from:
    #0 0x2b33574f71ff in __interceptor_malloc ../../.././libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x7532cb in hco_valinit_3d_dp /n/home05/msulprizio/RD/HEMCO_sanitize/CodeDir/src/Core/hco_arr_mod.F90:925
```
Is that covered by the issues in `ReadList_Init`?

**Comment by yantosca** [comment]
That is within `hco_arr_mod`.  Working on this one currently.

**Comment by yantosca** [comment]
I've gotten it down to 7.5MB:
```console
=================================================================
==9941==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 6226560 byte(s) in 5 object(s) allocated from:
    #0 0x2ba53e66e8ff in __interceptor_malloc /tmp/ryantosca/spack-stage/spack-stage-gcc-11.2.0-hbazyfn6aayq2jlpog7uck7w5xa5nmm4/spack-src/libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0xd0a670 in hco_valinit_3d_dp /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Core/hco_arr_mod.F90:982
    #2 0xd06b28 in __hco_arr_mod_MOD_hco_arrassert_3d_hp /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Core/hco_arr_mod.F90:1104
    #3 0x1119714 in __hco_geotools_mod_MOD_hco_calcvertgrid /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Core/hco_geotools_mod.F90:1216
    #4 0x422b96 in extstate_setfields /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Interfaces/Standalone/hcoi_standalone_mod.F90:2760
    #5 0x45cf47 in __hcoi_standalone_mod_MOD_hcoi_sa_run /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Interfaces/Standalone/hcoi_standalone_mod.F90:622
    #6 0x45fbe7 in __hcoi_standalone_mod_MOD_hcoi_standalone_run /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Interfaces/Standalone/hcoi_standalone_mod.F90:202
    #7 0x4054cd in hemco_standalone /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Interfaces/Standalone/hemco_standalone.F90:136
    #8 0x4059f6 in main /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Interfaces/Standalone/hemco_standalone.F90:21
    #9 0x2ba54134b554 in __libc_start_main (/lib64/libc.so.6+0x22554)

Direct leak of 1245312 byte(s) in 1 object(s) allocated from:
    #0 0x2ba53e66e8ff in __interceptor_malloc /tmp/ryantosca/spack-stage/spack-stage-gcc-11.2.0-hbazyfn6aayq2jlpog7uck7w5xa5nmm4/spack-src/libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0xd0a670 in hco_valinit_3d_dp /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Core/hco_arr_mod.F90:982
    #2 0xd06b28 in __hco_arr_mod_MOD_hco_arrassert_3d_hp /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Core/hco_arr_mod.F90:1104
    #3 0x10e6905 in __hco_geotools_mod_MOD_hco_calcvertgrid /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Core/hco_geotools_mod.F90:1076
    #4 0x422b96 in extstate_setfields /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Interfaces/Standalone/hcoi_standalone_mod.F90:2760
    #5 0x45cf47 in __hcoi_standalone_mod_MOD_hcoi_sa_run /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Interfaces/Standalone/hcoi_standalone_mod.F90:622
    #6 0x45fbe7 in __hcoi_standalone_mod_MOD_hcoi_standalone_run /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Interfaces/Standalone/hcoi_standalone_mod.F90:202
    #7 0x4054cd in hemco_standalone /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Interfaces/Standalone/hemco_standalone.F90:136
    #8 0x4059f6 in main /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Interfaces/Standalone/hemco_standalone.F90:21
    #9 0x2ba54134b554 in __libc_start_main (/lib64/libc.so.6+0x22554)

Direct leak of 25840 byte(s) in 34 object(s) allocated from:
    #0 0x2ba53e66e8ff in __interceptor_malloc /tmp/ryantosca/spack-stage/spack-stage-gcc-11.2.0-hbazyfn6aayq2jlpog7uck7w5xa5nmm4/spack-src/libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0xfc1226 in __hco_filedata_mod_MOD_filedata_init /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Core/hco_filedata_mod.F90:172
    #2 0xe831ea in config_readcont /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Core/hco_config_mod.F90:1160
    #3 0xe8afb5 in __hco_config_mod_MOD_config_readfile /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Core/hco_config_mod.F90:381
    #4 0xe781fb in config_readcont /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Core/hco_config_mod.F90:799
    #5 0xe8ab31 in __hco_config_mod_MOD_config_readfile /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Core/hco_config_mod.F90:364
    #6 0x45d851 in __hcoi_standalone_mod_MOD_hcoi_sa_init /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Interfaces/Standalone/hcoi_standalone_mod.F90:279
    #7 0x45fad6 in __hcoi_standalone_mod_MOD_hcoi_standalone_run /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Interfaces/Standalone/hcoi_standalone_mod.F90:194
    #8 0x4054cd in hemco_standalone /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Interfaces/Standalone/hemco_standalone.F90:136
    #9 0x4059f6 in main /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Interfaces/Standalone/hemco_standalone.F90:21
    #10 0x2ba54134b554 in __libc_start_main (/lib64/libc.so.6+0x22554)

Direct leak of 25080 byte(s) in 33 object(s) allocated from:
    #0 0x2ba53e66e8ff in __interceptor_malloc /tmp/ryantosca/spack-stage/spack-stage-gcc-11.2.0-hbazyfn6aayq2jlpog7uck7w5xa5nmm4/spack-src/libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0xfc1226 in __hco_filedata_mod_MOD_filedata_init /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Core/hco_filedata_mod.F90:172
    #2 0xe831ea in config_readcont /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Core/hco_config_mod.F90:1160
    #3 0xe8ab31 in __hco_config_mod_MOD_config_readfile /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Core/hco_config_mod.F90:364
    #4 0xe781fb in config_readcont /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Core/hco_config_mod.F90:799
    #5 0xe8ab31 in __hco_config_mod_MOD_config_readfile /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Core/hco_config_mod.F90:364
    #6 0x45d851 in __hcoi_standalone_mod_MOD_hcoi_sa_init /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Interfaces/Standalone/hcoi_standalone_mod.F90:279
    #7 0x45fad6 in __hcoi_standalone_mod_MOD_hcoi_standalone_run /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Interfaces/Standalone/hcoi_standalone_mod.F90:194
    #8 0x4054cd in hemco_standalone /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Interfaces/Standalone/hemco_standalone.F90:136
    #9 0x4059f6 in main /n/holyscratch01/jacob_lab/ryantosca/hco/HEMCO_sanitize/CodeDir/src/Interfaces/Standalone/hemco_standalone.F90:21
    #10 0x2ba54134b554 in __libc_start_main (/lib64/libc.so.6+0x22554)

SUMMARY: AddressSanitizer: 7522792 byte(s) leaked in 73 allocation(s).
[holyjacob01 HEMCO_sanitize]$ 
```
There is still some funkiness going on in `hco_arr_mod.F90`.  Will dig further.

**Comment by yantosca** [comment]
actually it 7MB for a 6hr test run.  For a 1-month run it could be a larger amount of memory leaked.

For the standalone it looks like the big memory leaks are coming from hco_geotools_mod, where it is reading the met fields to make the vertical pressure coordinages.  That would be PEDGE, TK, PSFC, ZFSC, and BXHEIGHT.  These get passed to HCO_ArrAssert, which calls HCO_ValInit. There I we have code such as:

```F90
    Val => NULL()
    Alloc = .FALSE.
    IF ( nx > 0 .and. ny > 0 .and. nz > 0 ) THEN
       ALLOCATE(Val(nx,ny,nz),STAT=AS)
       IF(AS/=0) THEN
          WRITE(*,*) 'Arr3D value allocation error'
          RC = HCO_FAIL
          IF ( ASSOCIATED( Val ) ) DEALLOCATE( Val )
          Val => NULL()
          RETURN
       ENDIF
       Val   = 0.0_dp
       Alloc = .TRUE.
    ENDIF
```
I think the nullification of Val (if Val is already allocated) may be causing the memory leak.   I will continue to look into this.

Also tagging @jimmielin

**Comment by yantosca** [comment]
Here is an update: 

1. I removed the largest memory leak (in `hco_geotools_mod.F90)`, in commit: 99de64e7a5c401dbd6e72e51622cd2abc46b27fb.  The issue was that the `HcoState%Grid%BXHEIGHT_M` was nullified even though it was still allocated.  The fix was to call Hco_ArrCleanup to deallocate the memory properly, thus removing the memory leak.

2. I removed the memory leaks in `hco_readlist_mod.F90`, see commit: 5ae8fde2885abf8aebb6937f46e53b66d9fa5d5b for details.

3. I removed the memory leaks in `hcox_dustdead_mod.F90` and `hcox_tomas_dustdead_mod.F90`, see commit: 220bcc60f39a4ce258e0d829e7178e59db72e0b0.  This was due to an a field of `Inst` that was allocated but never deallocated.

4. There are still memory leaks incurred in `hco_filedata_mod.F90` and `hco_config_mod.F90`.  It is because the `Dta` object is being repeatedly allocated in routine `HCO_FileData_Init`.  I have not yet found a solution to this, as it might involve substantial rewriting of the code.

Tagging @msulprizio @jimmielin @christophkeller

**Comment by yantosca** [comment]
I have also updated the code in `hco_arr_mod.F90` to make sure that errors now trigger an error message and return (instead of returning silently with no message).

**Comment by yantosca** [comment]
I have isolated the remaining memory leak to the list container in `hco_config_mod.F90`.  It occurs when we have entries of `HEMCO_Config.rc` that are to be read from the file listed above it, i.e
```kconfig
0 CEDS_CO_IND     $ROOT/CEDS/v2021-06/$YYYY/CO-em-anthro_CMIP_CEDS_$YYYY.nc  CO_ind  1750-2019/1-12/1/0 C xy kg/m2/s CO    26        1 5
0 CEDS_SOAP_IND   -                                                          -       -                  - -  -       SOAP  26/280    1 5
```
This is due to the local variable `Dta` being allocated but not finalized properly.  I have rewritten the loop in `hco_config_mod.F90` (see commit 2de6437bd78145ad7cd33c4d6f8c4bf00d2fdc08 )to try to avoid the memory leaks but have not yet been successful.  Still investigating.

**Comment by yantosca** [comment]
PR #194 has been merged into the HEMCO 3.6.1 development stream so we can close this issue.  We will look into the remaining memory leak at a later time.
